services/grafana/auth_server.go

```diff
@@ -108,13 +146,52 @@ func (s *AuthServer) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 			"message": err.message,
 		}
 		rw.Header().Set("Content-Type", "application/json")
-		rw.WriteHeader(runtime.HTTPStatusFromCode(err.code))
+
+		rw.WriteHeader(authenticationErrorCode)
 		if err := json.NewEncoder(rw).Encode(m); err != nil {
-			s.l.Warnf("%s", err)
+			l.Warnf("%s", err)
 		}
 	}
 }
 
+// mustSetup returns true if AWS instance ID must be checked.
+func (s *AuthServer) mustSetup(rw http.ResponseWriter, req *http.Request, l *logrus.Entry) bool {
+	// /setup page uses this API.
+	if req.URL.Path == "/server.Server/AWSInstanceCheck" || req.URL.Path == "/v1/AWSInstanceCheck" {
+		return false
+	}
+
+	// This header is used to pass information that setup is required from auth_request subrequest
+	// to normal request to return redirect with location - something that auth_request can't do.
+	const mustSetupHeader = "X-Must-Setup"
+
+	// Redirect to /setup page.
+	if req.Header.Get(mustSetupHeader) != "" {
+		const redirectCode = 303 // temporary, not cacheable, always GET
+		l.Warnf("AWS instance ID must be checked, returning %d with Location.", redirectCode)
+		rw.Header().Set("Location", "/setup")
+		rw.WriteHeader(redirectCode)
+		return true
+	}
+
+	// Use X-Test-Must-Setup header for testing.
+	// There is no way to skip check, only to enforce it.
+	mustCheck := s.checker.MustCheck()
+	if req.Header.Get("X-Test-Must-Setup") != "" {
+		l.Debug("X-Test-Must-Setup is present, enforcing AWS instance ID check.")
+		mustCheck = true
+	}
+
+	if mustCheck {
+		l.Warnf("AWS instance ID must be checked, returning %d with %s.", authenticationErrorCode, mustSetupHeader)
+		rw.Header().Set(mustSetupHeader, "1") // any non-empty value is ok
```

Comment:
@BupycHuk there

Reply:
So we add header to response and nginx sends second request with this header, right?

Reply:
Yes, that's the core of it. A ton of comments there and in nginx.conf explain it in more details.